File: backend/src/ai/genui/morning.py
"""genui/morning.py — the Morning Story, composed (LINE L1, VG-20).

The story is the Standup told in the pocket: one card per colleague, the
one who needs the owner first, quiet days last. VG-20 sketched "a daily
narrative generation job"; the assessment found the narrative already
existed — DRIVER's Standup composes it client-side from three shipped
reads — so this module **ports that composition server-side** (the audio
job needs it there) and keeps it a projection over the estate's truth.
The sentences here and in ``StandupSurface.tsx`` must tell the same story;
if the two ever drift, the phone and the desk disagree about yesterday,
and the fix is here.

The read path (``morning_story``): a stored row for today — written by
L2's job, its audio aligned to its cards — is served as-is, because the
story is *the morning's telling*, not a live view (the Standup is the
live view). No row → compose fresh, text-only, with the absence named
(``degraded_reason: "not_generated"``) rather than silent.
"""
from __future__ import annotations

import uuid
from datetime import date, datetime, timedelta
from typing import Any, Iterable

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.ai.genui.estate import estate_view
from src.ai.genui.trays import _pending_with_entities
from src.ai.orm.execution import ExecutionRun

__all__ = ["compose_morning_story", "morning_story"]


def compose_morning_story(
    estate: dict[str, Any],
    runs: Iterable[tuple[str, str]],
    waiting_entity_ids: set[str],
    now: datetime,
) -> list[dict[str, Any]]:
    """One card per colleague. Pure — the Standup's composition, ported.

    ``runs`` are ``(entity_id, status)`` pairs already bounded to the last
    day; the counting lives here so the job and the endpoint cannot count
    differently.
    """
    completed: dict[str, int] = {}
    failed: dict[str, int] = {}
    running: dict[str, int] = {}
    for entity_id, status in runs:
        if status == "COMPLETED":
            completed[entity_id] = completed.get(entity_id, 0) + 1
        elif status == "FAILED":
            failed[entity_id] = failed.get(entity_id, 0) + 1
        elif status == "RUNNING":
            running[entity_id] = running.get(entity_id, 0) + 1

    cards: list[dict[str, Any]] = []
    for district in estate.get("districts", []):
        for colleague in district.get("colleagues", []):
            entity_id = str(colleague.get("entity_id"))
            sentences: list[str] = []
            done = completed.get(entity_id, 0)
            if done == 1:
                sentences.append("Finished one piece of work since yesterday.")
            elif done > 1:
                sentences.append(
                    f"Finished {done} pieces of work since yesterday.")
            wrong = failed.get(entity_id, 0)
            if wrong == 1:
                sentences.append("One thing went wrong — it is in the trace.")
            elif wrong > 1:
                sentences.append(
                    f"{wrong} things went wrong — they are in the traces.")
            if running.get(entity_id, 0) > 0:
                sentences.append("Is working on something right now.")
            waiting = entity_id in waiting_entity_ids
            if waiting:
                sentences.append("Is waiting on you.")
            if not sentences:
                sentences.append("A quiet day — nothing to report.")
            cards.append({
                "entity_id": entity_id,
                "name": colleague.get("name"),
                "district": district.get("process_code"),
                "sentences": sentences,
                "waiting": waiting,
                # The job fills this; a fresh composition honestly has none.
                "audio": None,
            })
    # Whoever needs the owner comes first; quiet days come last.
    cards.sort(key=lambda card: (not card["waiting"]))
    return cards


async def _runs_since_yesterday(
    db: AsyncSession, company_id: uuid.UUID, now: datetime,
) -> list[tuple[str, str]]:
    rows = (
        await db.execute(
            select(ExecutionRun.entity_id, ExecutionRun.status)
            .where(
                ExecutionRun.company_id == company_id,
                ExecutionRun.created_at >= now - timedelta(days=1),
            ))
    ).all()
    return [(str(entity_id), str(status)) for entity_id, status in rows]


async def compose_for_company(
    db: AsyncSession, company_id: uuid.UUID, *, now: datetime,
) -> list[dict[str, Any]]:
    """The three shipped reads, then the pure composition — shared by the
    endpoint's fresh path and L2's job so they cannot drift."""
    estate = await estate_view(db, company_id, now=now)
    runs = await _runs_since_yesterday(db, company_id, now)
    waiting = {
        str(entity_id)
        for _approval, entity_id, _name in await _pending_with_entities(
            db, company_id)
    }
    return compose_morning_story(estate, runs, waiting, now)


async def morning_story(
    db: AsyncSession, company_id: uuid.UUID, *, now: datetime | None = None,
) -> dict[str, Any]:
    """Today's story for the Line."""
    now = now or datetime.utcnow()
    story_date: date = now.date()

    from src.ai.genui.models import MorningStory

    row = await db.get(MorningStory, (company_id, story_date))
    if row is not None:
        return {
            "story_date": story_date.isoformat(),
            "cards": row.cards,
            "generated_at": row.generated_at.isoformat(),
            "degraded_reason": row.degraded_reason,
        }
    cards = await compose_for_company(db, company_id, now=now)
    return {
        "story_date": story_date.isoformat(),
        "cards": cards,
        "generated_at": None,
        "degraded_reason": "not_generated",
    }

File: vihara/src/fixtures/morning.ts
/**
 * The Morning Story's data, shaped to `GET /ai/genui/line/morning` (LINE L1).
 *
 * The endpoint is `backend/src/ai/genui/morning.py`; the row it usually serves is
 * written at 02:25 UTC by `morning_job.py`. Field names are camel-cased to the
 * house convention and map one-to-one onto the wire's snake_case, so R-4 swaps
 * the source and not the surface (`story_date`, `entity_id`, `data_b64`,
 * `degraded_reason`).
 *
 * Two properties of this fixture are load-bearing rather than decorative.
 *
 * **The sentences are only the six the server can compose.** `compose_morning_story`
 * writes counts, not narrative — "Finished N pieces of work since yesterday.",
 * "One thing went wrong — it is in the trace.", "Is working on something right
 * now.", "Is waiting on you.", "A quiet day — nothing to report." — and nothing
 * else. Writing richer prose here would design a surface against a composition
 * the platform does not perform, which is the same class of error as inventing a
 * number. The counts themselves agree with the desk's `STANDUP` facts
 * (`fixtures/decisions.ts`) and the two waiting cards agree with
 * `STILL.handsRaised`, because morning.py's own docstring says it: if the phone
 * and the desk drift, they disagree about yesterday.
 *
 * **The degradation is the mid-story one.** `_synthesize` fills clips in order
 * and returns `"tts_failed"` on the first exception, keeping what it already
 * made — so a real broken morning is two voiced cards and three text ones, not a
 * silent story. That is the shape most worth designing against, because it is
 * the one where a surface can quietly pretend nothing is missing.
 */

/** Gemini's native rate, mirrored from `morning_job.WAV_SAMPLE_RATE`. A wrong
 *  value here does not raise — it plays at the wrong speed. */
const SAMPLE_RATE = 24000;

export interface MorningAudio {
  mime: string;
  /** Base64 of a RIFF/WAVE payload — `morning_job.wav_wrap` around 16-bit mono PCM. */
  dataB64: string;
}

export interface MorningCard {
  entityId: string;
  name: string;
  /** The district's process code, which is all the endpoint sends: `P08`. */
  district: string;
  /** Her telling, one sentence per composed fact. The clip says these joined. */
  sentences: string[];
  waiting: boolean;
  /** `null` where the job never made a clip for this card. Never a broken src. */
  audio: MorningAudio | null;
}

/** The four the job can record. A row may also be clean and a card still silent
 *  — see `MorningStorySurface`, which has a sentence for that too. */
export type DegradedReason = "wallet" | "not_configured" | "tts_failed" | "not_generated";

export interface MorningStory {
  /** A calendar date, not an instant. */
  storyDate: string;
  cards: MorningCard[];
  /** When the telling was made. `null` means the job has not run and this was
   *  composed on read — a different morning, honestly labelled. */
  generatedAt: string | null;
  degradedReason: DegradedReason | null;
}

/* ------------------------------------------------------------------ the clip
   A stand-in for her voice until R-4 wires the endpoint.

   It is synthesised rather than pasted in as base64 for two reasons. A real WAV
   is ~50 KB per card, and a literal that size in a source file is fifty
   kilobytes nobody will ever read again. More importantly, the alternative was a
   silent or absent `src` — and a play button that does nothing is exactly the
   lie this surface exists to refuse. The bytes are made the same way the job
   makes them, so the client's reader and the server's writer agree on the
   format before either is wired. */

function base64(bytes: Uint8Array): string {
  let binary = "";
  // 8 KB at a time: spreading a 50 KB array into fromCharCode overflows the
  // argument stack on some engines, and it does it silently until it does not.
  const CHUNK = 0x2000;
  for (let i = 0; i < bytes.length; i += CHUNK) {
    binary += String.fromCharCode(...bytes.subarray(i, i + CHUNK));
  }
  return btoa(binary);
}

/** The RIFF/WAVE header, byte for byte as `morning_job.wav_wrap` writes it. */
function wav(pcm: Int16Array): string {
  const bytes = new Uint8Array(44 + pcm.length * 2);
  const view = new DataView(bytes.buffer);
  const tag = (at: number, s: string) => {
    for (let i = 0; i < s.length; i++) view.setUint8(at + i, s.charCodeAt(i));
  };
  tag(0, "RIFF");
  view.setUint32(4, 36 + pcm.length * 2, true);
  tag(8, "WAVE");
  tag(12, "fmt ");
  view.setUint32(16, 16, true); // subchunk size
  view.setUint16(20, 1, true); // PCM
  view.setUint16(22, 1, true); // mono
  view.setUint32(24, SAMPLE_RATE, true);
  view.setUint32(28, SAMPLE_RATE * 2, true); // byte rate
  view.setUint16(32, 2, true); // block align
  view.setUint16(34, 16, true); // bits per sample
  tag(36, "data");
  view.setUint32(40, pcm.length * 2, true);
  for (let i = 0; i < pcm.length; i++) view.setInt16(44 + i * 2, pcm[i]!, true);
  return base64(bytes);
}

/** A short warm tone, low and falling, with a soft attack and a long release —
 *  a placeholder that reads as a voice note rather than as an alert chime. */
function standInClip(step: number): MorningAudio {
  const seconds = 0.8;
  const n = Math.round(SAMPLE_RATE * seconds);
  const pcm = new Int16Array(n);
  const base = 132 + step * 9;
  for (let i = 0; i < n; i++) {
    const t = i / SAMPLE_RATE;
    const env = Math.min(1, t / 0.07) * Math.min(1, (seconds - t) / 0.3);
    const hz = base * (1 - t * 0.06);
    const wave =
      Math.sin(2 * Math.PI * hz * t) * 0.62 + Math.sin(4 * Math.PI * hz * t) * 0.2;
    pcm[i] = Math.round(wave * env * 8200);
  }
  return { mime: "audio/wav", dataB64: wav(pcm) };
}

/* ----------------------------------------------------------------- the story
   Thursday's morning, told at 02:25 UTC, with her voice lost after the second
   card. Whoever needs the owner comes first and quiet days come last, which is
   `compose_morning_story`'s sort and not a layout preference. */

export const MORNING: MorningStory = {
  storyDate: "2026-07-30",
  generatedAt: "2026-07-30T02:25:00.000Z",
  degradedReason: "tts_failed",
  cards: [
    {
      entityId: "AGT-046",
      name: "Meera",
      district: "P08",
      sentences: [
        "Finished 33 pieces of work since yesterday.",
        "One thing went wrong — it is in the trace.",
        "Is working on something right now.",
        "Is waiting on you.",
      ],
      waiting: true,
      audio: standInClip(0),
    },
    {
      entityId: "AGT-041",
      name: "Anjali",
      district: "P08",
      sentences: [
        "Finished 9 pieces of work since yesterday.",
        "Is waiting on you.",
      ],
      waiting: true,
      audio: standInClip(1),
    },
    {
      // The clip for this card is where synthesis raised. Everything from here
      // is text, and the surface says so on each card rather than once at the top.
      entityId: "AGT-038",
      name: "Ravi",
      district: "P08",
      sentences: [
        "Finished 16 pieces of work since yesterday.",
        "2 things went wrong — they are in the traces.",
        "Is working on something right now.",
      ],
      waiting: false,
      audio: null,
    },
    {
      // A colleague with nothing to report. The server has one sentence for
      // that and this is it — a thin card, honestly thin.
      entityId: "AGT-013",
      name: "Devika",
      district: "P03",
      sentences: ["A quiet day — nothing to report."],
      waiting: false,
      audio: null,
    },
    {
      entityId: "AGT-092",
      name: "Farhan",
      district: "P14",
      sentences: [
        "Finished 27 pieces of work since yesterday.",
        "Is working on something right now.",
      ],
      waiting: false,
      audio: null,
    },
  ],
};
